Target and Count encodings for categorical features

shiyu1994 opened the pull request:
This pull request is to support converting categorical features into CTR and count values. The CTR values are calculated by dividing the training data into folds, with a cross-validation style. 

The performance evaluation has been added in https://github.com/microsoft/LightGBM/pull/3234#issuecomment-691792107. Note that this is a basic version without ensemble trick. We will have a separate PR to implement ensemble trick to boost the performance of this categorical encoding method.

Detailed descriptions and guideline for reviews can be found below.

## Description
### Idea
LightGBM can handle categorical features internally, without any manual encoding into numerical values by users. The current approach that LightGBM handles categorical features can be found here https://lightgbm.readthedocs.io/en/latest/Features.html#optimal-split-for-categorical-features.
However, this approach has two known drawbacks:
1. Strong regularization is often required for this approach to be effective when the number of categories are high. The 3 hyperparameters `cat_l2` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#cat_l2), `cat_smooth` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#cat_smooth) and `max_cat_to_onehot` (https://lightgbm.readthedocs.io/en/latest/Parameters.html#max_cat_to_onehot) need to be tuned carefully.
2. The current approach requires sorting over call categories for every categorical feature on every node, which can be slow with a large number of categories.

Inspired by CatBoost, we implement two new approach for internal encoding of categorical features:
1. Encode with the categories by target (label) values (Unknown as target encoding). A category value `c` of feature `j` is encoded as
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\sum_i&space;I[x_{ij}=c]&space;y_i&space;&plus;&space;w&space;\cdot&space;p}{\sum_i&space;I[x_{ij}=c]&space;&plus;&space;w}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\sum_i&space;I[x_{ij}=c]&space;y_i&space;&plus;&space;w&space;\cdot&space;p}{\sum_i&space;I[x_{ij}=c]&space;&plus;&space;w}" title="\frac{\sum_i I[x_{ij}=c] y_i + w \cdot p}{\sum_i I[x_{ij}=c] + w}" /></a>
where <a href="https://www.codecogs.com/eqnedit.php?latex=x_{ij}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?x_{ij}" title="x_{ij}" /></a> is the value of feature `j` for data `i` in the training set, <a href="https://www.codecogs.com/eqnedit.php?latex=I" target="_blank"><img src="https://latex.codecogs.com/gif.latex?I" title="I" /></a> is the indicator function, <a href="https://www.codecogs.com/eqnedit.php?latex=y_i" target="_blank"><img src="https://latex.codecogs.com/gif.latex?y_i" title="y_i" /></a> is the label for data `i`, `p` is the prior value (which is the mean of labels over the whole dataset, by default), and `w` is the weight for the prior value.
2. Encode the categories by the number of appearances in the whole training data, i.e.,
<a href="https://www.codecogs.com/eqnedit.php?latex=\sum_i&space;I[x_{ij}=c]" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\sum_i&space;I[x_{ij}=c]" title="\sum_i I[x_{ij}=c]" /></a>

For better generalization ability, for the target encoding, we use a cross-validation style approach. The training data is first divided into `K` folds. When encoding the category values in the fold `k`, we only use the other `K-1` except fold `k` to calculate the encoded value. Thus, for each category `c`, we will have `K` different encoded values for a target encoding approach.

### New Parameters
We add a new parameter `category_encoders` to specify the encoding methods for categorical features. Users can specify multiple encoding methods at the same time, and separate them by commas. For example, with `category_encoders=target:0.5,count,target`, we'll use 3 encoding methods at the same time: target encoding with prior value `p=0.5`, count encoding, and target encoding with default prior value. Each encoding method creates a new feature for each categorical feature for training. We also allow a `raw` encoding method which indicates the current approach of handling categorical feature of LightGBM.

Besides, we also add two parameters `prior_weight` and `num_target_encoding_folds` to allow users specify the `p` and `K` in the target encoding.

## Implementation
The core of this PR is a new class `CategoryEncodingProvider`, which is defined in https://github.com/shiyu1994/LightGBM/blob/ctr/include/LightGBM/category_encoding_provider.hpp and https://github.com/shiyu1994/LightGBM/blob/ctr/src/io/category_encoding_provider.cpp. These two files amount to 2/3 of the code changes in the PR.
`CategoryEncodingProvider` works by 2 steps:
1. Takes the raw data as input (can be LibSVM file, numpy/pandas/scipy.sparse matrices), and calculates the encoding values. Store these encoding values in side its own members.
2. Expanding the raw data with an additional new feature for each categorical feature and each encoding method, **before** the data is fed into any internal data preprocessing modules by LightGBM (including bin finding, exclusive feature bundling and so on).

Since we need to support multiple data input formats, `CategoryEncodingProvider` is used in corresponding functions in `src/c_api.cpp` which is called when processing inputs of these formats. Including
1. For LibSVM files: `LGBM_DatasetCreateFromFile`.
2. For numpy/pandas matrix: `LGBM_DatasetCreateFromMat` and `LGBM_DatasetCreateFromMats`.
3. For row-wise sparse matrix: `LGBM_DatasetCreateFromCSR`.
4. For col-wise sparse matrix: `LGBM_DatasetCreateFromCSC`.

guolinke commented on include/LightGBM/ctr_provider.hpp:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  inline double EncodeCatValueForValidation(const int fid, const double feature_value) const {
+    const auto& ctr_encoding = ctr_encodings_.at(fid)[config_.num_ctr_folds];
```
avoid to use `.at()`, which is slow.

guolinke commented on include/LightGBM/config.h:
```diff
@@ -961,6 +961,19 @@ struct Config {
+  // desc = ``0`` means use the original categorical feature values
+  // desc = ``1`` means convert the categorical values into ctr values
+  // desc = ``2`` means using the count values and ctr values
+  int use_ctr = 0;
```
`count` only option?

guolinke commented on include/LightGBM/ctr_provider.hpp:
```diff
@@ -0,0 +1,332 @@
+    }
+    categorical_features.clear();
+    GenTrainingDataFoldID();
+    if(config_.use_ctr == 2) {
```
maybe pre-define the names of these CTR types, instead of using magic numbers.

guolinke commented on include/LightGBM/c_api.h:
```diff
@@ -206,7 +209,9 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetCreateFromCSC(const void* col_ptr,
                                                 const void* data,
+                                                const void* label,
                                                 int data_type,
+                                                int label_type,
```
I remember we fix the label type to `double`. Maybe using the dynamic type is not needed.

guolinke replied:
label type to double directly?

shiyu1994 replied:
The native floating point number of R only supports 64-bit, so I think keep `label_type` is necessary to support both R and python interfaces.

guolinke commented on src/c_api.cpp:
```diff
@@ -1042,21 +1084,27 @@ int LGBM_DatasetCreateFromMats(int32_t nmat,
       for (size_t k = 0; k < row.size(); ++k) {
-        if (std::fabs(row[k]) > kZeroThreshold || std::isnan(row[k])) {
+        // if ctr is used, zero categorical values should not be ignored
+        if (std::fabs(row[k]) > kZeroThreshold || std::isnan(row[k]) || (config.use_ctr > 0 && is_categorical_feature[k])) {
```
is CTR calculated by full dataset or sample set? sampled label is needed?

shiyu1994 replied:
It is calculated by full dataset. I will remove the sample label stuff.

jameslamb commented:
At your earliest convenience, can you please merge in `master`?  Some changes have been made to our CI jobs. I apologize for the inconvenience.

guolinke commented:
@shiyu1994 can you fix the conflict?

jameslamb reviewed (changes requested):
@shiyu1994 for the R side, I see you have this error in CI:

![image](https://user-images.githubusercontent.com/7608904/92498824-b4597d80-f1c0-11ea-8195-aa3fd0dbb546.png)

Can you please update https://github.com/microsoft/LightGBM/blob/82e2ff7a018f2466f8102c80a7bafb9324238d19/R-package/src/lightgbm_R.cpp#L681? Now that you've added two more arguments to this function, it should be

```cpp
  {"LGBM_DatasetCreateFromCSC_R"      , (DL_FUNC) &LGBM_DatasetCreateFromCSC_R      , 12},
```

guolinke commented on include/LightGBM/boosting.h:
```diff
@@ -131,10 +131,10 @@ class LIGHTGBM_EXPORT Boosting {
                           const PredictionEarlyStopInstance* early_stop) const = 0;
 
-  virtual void PredictRawByMap(const std::unordered_map<int, double>& features, double* output,
+  virtual void PredictRawByMap(std::unordered_map<int, double>& features, double* output,
```
Is that possible to update feature values before calling prediction? like have a function like GBDT::PrepraeFeatureValues, and call before prediction.

guolinke commented on include/LightGBM/c_api.h:
```diff
@@ -389,6 +398,15 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetGetNumData(DatasetHandle handle,
+ * \param[out] out The address to hold number of features
+ * \return 0 when succeed, -1 when failure happens
+ */
+LIGHTGBM_C_EXPORT int LGBM_DatasetGetNumOriginalFeature(DatasetHandle handle,
```
I remember that LGBM_DatasetGetNumFeature returns the number of original features. Maybe this additional api is not needed?

guolinke replied:
cc @shiyu1994

shiyu1994 replied:
Because the CTR construction is before dataset construction, the LGBM_DatasetGetNumFeature returns the number of features with CTR transformations. So I think this is necessary.

guolinke commented on include/LightGBM/config.h:
```diff
@@ -970,6 +970,26 @@ struct Config {
+
+  // desc = whether to keep the original feature values after the dataset is constructed
+  // desc = if set false, then once dataset is constructed, the cat_converters cannot be changed through parameters in train method
+  bool keep_raw_cat_data = false;
```
keep_raw_cat_data can be removed in this version? we can refine these together with linear and other dataset parameters.

guolinke commented on include/LightGBM/dataset_loader.h:
```diff
@@ -39,6 +41,14 @@ class DatasetLoader {
 
+  std::unordered_set<int>& GetParsedCategoricalFeatures() {
+    return categorical_features_;
+  }
```
return not const reference?

guolinke commented on include/LightGBM/dataset_loader.h:
```diff
@@ -48,11 +58,19 @@ class DatasetLoader {
+
+  std::vector<std::string> SampleTextDataFromFile(const char* filename, const Metadata& metadata, int rank, 
+    int num_machines, int* num_global_data, std::vector<data_size_t>* used_data_indices, 
+    std::vector<data_size_t>& sampled_data_indices);
```
why sampled_data_indices is needed? and pass reference?

guolinke commented on include/LightGBM/utils/text_reader.h:
```diff
@@ -170,18 +170,21 @@ class TextReader {
 
-  INDEX_T SampleFromFile(Random* random, INDEX_T sample_cnt, std::vector<std::string>* out_sampled_data) {
+  INDEX_T SampleFromFile(Random* random, INDEX_T sample_cnt, std::vector<std::string>* out_sampled_data,
+    std::vector<INDEX_T>& sampled_data_indices) {
```
only use `T*` or `const T&` in arguments.

guolinke commented on src/main.cpp:
```diff
@@ -5,13 +5,23 @@
+    }
+    conf_path[conf_path_string.size()] = '\0';
+    argv[1] = conf_path;
+    LightGBM::Application app(2, argv);
```
remember to remove these debug codes.

shiyu1994 commented:
Comparison between CTR and old categorical feature approach:
We use 7 datasets with categorical features and numerical features. We tune the hyperparameters of LightGBM with HyperOpt. 
The following table lists the range of tuning for parameters shared by CTR version and original version.

Parameter | Range
---|---
learning_rate | [e^-7, 1]
num_leavs | [e^1, e^7]
max_bin | [2^4, 2^10]
feature_fraction | [0.5, 1]
bagging_fraction | [0.5, 1]
min_data_in_leaf | [1, e^6]
min_sum_hessians_in_leaf | [e^-16, e^5]
lambda_l1 | [0, e^2]
lambda_l2 | [0, e^2]

The hyper-parameters of old categorical method are
Parameter | Range 
---|---
cat_smooth | [1,e^8]
cat_l2 | [0, e^6]
max_cat_threshold | [0, max_cat_count / 2]

where `max_cat_count` is the maximum number of categorical values of categorical features in the training set.

For CTR version, we fix the number of folds for CTR calculation to 4. And try both CTR with count and CTR only.

We use 5-fold cross validation. For the number of trials allowed to tune hyperparameters, we allow the old categorical method to use 500 trials. And for each of CTR with count and CTR only, we use 200 trials. The tuning for each algorithm and each dataset is repeated for 5 times (with 5 different CV fold partitions).

Following table shows the AUC on test sets

Method | Adult | Amazon | Appetency | Click | Internet | Kick | Upselling 
---|---|---|---|---|---|---|---
Old Categorical | **0.929203**<br>+/-0.000331|0.856392<br>+/-0.000738|0.841982<br>+/-0.002691|0.720245<br>+/-0.000060|0.959978<br>+/-0.000445|**0.665910**<br>+/-0.003899|0.864470<br>+/-0.000858
CTR and Count |0.929112<br>+/-0.000394|**0.863328**<br>+/-0.001924|0.850926<br>+/-0.001570|**0.741018**<br>+/-0.000200|**0.960743**<br>+/-0.000411|0.644784<br>+/-0.009142|**0.864755**<br>+/-0.000763
CTR only |0.929004<br>+/-0.000435|0.852688<br>+/-0.002328|**0.853207**<br>+/-0.001834|0.735202<br>+/-0.000144|TBA|0.656863<br>+/-0.002550|0.864416<br>+/-0.000881

guolinke commented:
@shiyu1994 can you update the latest results, and fix the conflicts as well?

gzls90 commented:
Sounds intriguing feature, just few questions as digging deeper into code is bit too time consuming:
1) Is the idea is based on (or similar to) Catboost's categorical handling?
2) If so, then will it also support multiple feature combination mapping (that is CTR of group of categorical features instead of single one)?
3) If 1) and 2) true, then i have experience of Catboost having explosive memory usage due to categorical features handling that way. Would it be prevented in this PR in early implementation stages (or solutions how to handle it) to make it efficient and usable for large datasets and model compactness?

shiyu1994 commented:
> Sounds intriguing feature, just few questions as digging deeper into code is bit too time consuming:
> 
> 1. Is the idea is based on (or similar to) Catboost's categorical handling?
> 2. If so, then will it also support multiple feature combination mapping (that is CTR of group of categorical features instead of single one)?
> 3. If 1) and 2) true, then i have experience of Catboost having explosive memory usage due to categorical features handling that way. Would it be prevented in this PR in early implementation stages (or solutions how to handle it) to make it efficient and usable for large datasets and model compactness?

1. The idea is similar. The difference is that we do not use random permutation to calculate the CTR value. Instead, we randomly divide the training data into folds, and calculate the CTR values in a cross-validation manner.
2. No. We don't support combination of categorical features in this PR.

guolinke commented:
@shiyu1994 can you fix the conflicts?

guolinke commented on docs/Parameters.rst:
```diff
@@ -1184,6 +1184,31 @@ GPU Parameters
+
+-  ``prior_weight`` :raw-html:`<a id="prior_weight" title="Permalink to this parameter" href="#prior_weight">&#x1F517;&#xFE0E;</a>`, default = ``1.0f``, type = double
+
+   -  weight of prior in CTR calculation
```
add an equation about how `prior` and `prior_weight` are used in CTR calculation?

guolinke commented on R-package/src/lightgbm_R.cpp:
```diff
@@ -107,7 +107,7 @@ LGBM_SE LGBM_DatasetCreateFromMat_R(LGBM_SE data,
   double* p_mat = R_REAL_PTR(data);
   DatasetHandle handle = nullptr;
-  CHECK_CALL(LGBM_DatasetCreateFromMat(p_mat, C_API_DTYPE_FLOAT64, nrow, ncol, COL_MAJOR,
+  CHECK_CALL(LGBM_DatasetCreateFromMat(p_mat, nullptr, C_API_DTYPE_FLOAT64, C_API_DTYPE_NONE, nrow, ncol, COL_MAJOR,
```
R package does not support?

guolinke commented on src/io/ctr_provider.cpp:
```diff
@@ -0,0 +1,622 @@
+  }
+  return cat_converter->CalcValue(label_sum, total_count, total_count);
+}
+
```
Is that possible to use template to merge these `ConvertCatToCTR` ?

guolinke commented on src/io/dataset_loader.cpp:
```diff
@@ -222,7 +240,14 @@ Dataset* DatasetLoader::LoadFromFile(const char* filename, int rank, int num_mac
+      } else {
+        sample_data = SampleTextDataFromFileWithIndices(filename, dataset->metadata_, rank,
+          num_machines, &num_global_data, &used_data_indices, &sampled_indices);
+      }
```
move the `if (ctr_provider == nullptr) ... else ...` into SampleTextDataFromFile, then decide which function to call? to reduce duplicate codes.

guolinke commented on src/io/dataset_loader.cpp:
```diff
@@ -947,17 +1028,79 @@ std::vector<std::string> DatasetLoader::SampleTextDataFromFile(const char* filen
   return out_data;
 }
 
+std::vector<std::string> DatasetLoader::SampleTextDataFromFileWithIndices(const char* filename, const Metadata& metadata,
```
merge with `SampleTextDataFromFile` by template?

guolinke commented on src/io/dataset_loader.cpp:
```diff
@@ -1303,6 +1444,7 @@ void DatasetLoader::ExtractFeaturesFromFile(const char* filename, const Parser*
           }
         }
       }
+      // TODO(shiyu1994): should be start_idx + i ?
```
still todo?

shiyu1994 replied:
This should be `start_idx + i`, and it has been a potential bug since very long time ago. In line 1384, we use `PushData(tid, sub_feature, start_idx + i...` in the previous few lines. So using `dataset->FinishOneRow(tid, i,` here is an obvious inconsistence. The true line index is `start + i`, instead of `i`.

This code is not changed in PR, it is inherited from very early version. And it has long been a potential bug.

guolinke commented on R-package/R/lgb.Dataset.R:
```diff
@@ -206,19 +206,29 @@ Dataset <- R6::R6Class(
+            label <- NULL
+          } else {
+            label <- self$getinfo(name = "label")
+          }
```
remove label type in R

guolinke commented on R-package/src/lightgbm_R.cpp:
```diff
@@ -106,8 +110,10 @@ LGBM_SE LGBM_DatasetCreateFromMat_R(LGBM_SE data,
+  double* p_label = R_REAL_PTR(label);
   DatasetHandle handle = nullptr;
-  CHECK_CALL(LGBM_DatasetCreateFromMat(p_mat, C_API_DTYPE_FLOAT64, nrow, ncol, COL_MAJOR,
+  int label_type = p_label == nullptr ? C_API_DTYPE_NONE : C_API_DTYPE_FLOAT64;
```
remove label type

guolinke commented on include/LightGBM/c_api.h:
```diff
@@ -28,6 +28,7 @@ typedef void* FastConfigHandle; /*!< \brief Handle of FastConfig. */
 #define C_API_DTYPE_FLOAT64 (1)  /*!< \brief float64 (double precision float). */
 #define C_API_DTYPE_INT32   (2)  /*!< \brief int32. */
 #define C_API_DTYPE_INT64   (3)  /*!< \brief int64. */
+#define C_API_DTYPE_NONE   (4)  /*!< \brief None. */
```
remove this

guolinke commented on include/LightGBM/c_api.h:
```diff
@@ -148,7 +149,10 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetPushRowsByCSR(DatasetHandle dataset,
  * \param data Pointer to the data space
+ * \param label Pointer to the label space
  * \param data_type Type of ``data`` pointer, can be ``C_API_DTYPE_FLOAT32`` or ``C_API_DTYPE_FLOAT64``
+ * \param label_type Type of ``data`` pointer, can be ``C_API_DTYPE_FLOAT32`` or ``C_API_DTYPE_FLOAT64``
```
remove type

guolinke commented:
Thank you @shiyu1994 , finally, we can merge this PR. 
The next step is to remove the old categorical features supports and update the documentations accordingly.

guolinke commented:
@jameslamb can you check the R's part?

StrikerRUS commented:
Please let me take a look at Python part.

I see this PR breaks bakward compatibility with a lot of C API functions for Dataset creation. Does it mean next release will be `4.0.0`?

guolinke commented:
Hi @StrikerRUS , yeah, we may have 4.0.0 release, there are sereval on-going works need to in 4.0.0 as well.
I think we can release a 3.2.0 stable release recently, then merge this PR.

jameslamb commented:
/gha run r-valgrind

jameslamb commented on docs/Parameters.rst:
```diff
@@ -1182,6 +1182,49 @@ GPU Parameters
 
    -  **Note**: can be used only in CUDA implementation
 
+CTR Parameters
+--------------
```
I can review this more thoroughly tomorrow, but I have a fundamental question...what does "CTR" stand for? Could https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html#categorical-feature-support or some other section of the documentation be expanded to describe this? If I read this parameter document, I would not know what CTR means, so I wouldn't know what it means to set a prior, for example.

guolinke replied:
cc @shiyu1994

StrikerRUS commented:
@guolinke 
> there are sereval on-going works need to in 4.0.0 as well.

Sounds intriguingly! 🙂 

> I think we can release a 3.2.0 stable release recently, then merge this PR.

Nice plan!

> The next step is to remove the old categorical features supports and update the documentations accordingly.

Can't two methods of categorical features support live together so that users will have a choice?

shiyu1994 commented:
Currently the old categorical features is kept (users can enable it by add `"raw"` in the `cat_converters` parameter). But if the new approach is good enough, it would be better to remove the old one. Since the old method requires sorting categorical values in every leaf, which can be expensive. Also, it often requires very strong regularization to prevent overfitting, the parameter tuning process is more difficult then ordinary encoding methods.

StrikerRUS commented:
@shiyu1994 Thank you very much for the explanation!

Does old method require a lot of maintenance burden? If it will be easy to just leave old method in the current state, maybe we can raise warning for `raw` value of `cat_converters` parameter with your explanations but still allow users access it?

According to https://github.com/microsoft/LightGBM/pull/3234#issuecomment-691792107, old method can outperform new ones in some cases.

StrikerRUS commented:
@shiyu1994 Please sync with `master` to check whether CUDA version works with new features.

StrikerRUS commented:
@shiyu1994 
> The experiments is not finished yet.

Sorry, does it mean that this PR is still `WIP`? Or changes (if any) will be made in a follow-up PR?

shiyu1994 commented:
Allow both master and ctr branches to have at most 200 trials per dataset. And allow ctr branches to search for using either `cat_converters=ctr,count` or `cat_converters=ctr`. Unfortunately, ctr branch now shows advantage in only 3 out of 8 datasets. I will continue to optimize it to see if this can be improved.

|Dataset | CTR AUC | Master AUC |
| --- | --- | --- |
|adult| 0.929310±0.000344 |  **0.929336**±0.000287 |
|amazon| **0.858445**±0.003533 |  0.854134±0.002725 |
|appetency| **0.849800**±0.002979 |  0.838189±0.004230 |
|click| **0.740799**±0.000269 |  0.720182±0.000323 |
|internet| 0.959814±0.001270 |  **0.959849**±0.000314 |
|kick| 0.643412±0.006572 |  **0.663514**±0.003772 |
|upselling| 0.862846±0.001735 |  **0.863376**±0.001305 |
|nips_b| 0.615565±0.006900 |  **0.617608**±0.002437 |

In the previous version https://github.com/microsoft/LightGBM/pull/3234#issuecomment-691792107, we allow master to use 500 trials, and ctr branch to use 200 for `cat_converters=ctr` and 200 for `cat_converters=ctr,count`, and ctr branch wins 5 out of 7. I think perhaps fixing `cat_converters` for a dataset would be better, since whether `count` is useful or not often depends on the dataset, not other hyperparameters. So trying to searching `cat_converters` in `{"ctr", "ctr,count"}` may not be a good idea.

shiyu1994 commented:
@StrikerRUS @jameslamb Could you help to check the Python and R part of this PR, we are ready to merge this. We still keep the old category handling method of LightGBM in this PR, because we did not observe a consistence advantage of our new method in all datasets. The documentation has been updated to include the new categorical encoding methods. And we've renamed the `ctr` stuff to more comprehensible `target encoding`.
But we've figure out a method to improve the performance of current encoding methods, which requires more changes and will be reserved for another PR.

StrikerRUS commented:
@shiyu1994 Sure, will do this soon!

jameslamb commented:
ok sure! I'll look in the next day or two

StrikerRUS commented on include/LightGBM/c_api.h:
```diff
@@ -389,6 +397,15 @@ LIGHTGBM_C_EXPORT int LGBM_DatasetGetNumData(DatasetHandle handle,
                                                 int* out);
 
+/*!
+ * \brief Get number of original features.
```
```suggestion
 * \brief Get number of original features (before converting categorical features).
```

StrikerRUS commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,661 @@
+  // dump a dictionary to string
+  template <typename T>
+  static std::string DumpDictToString(const std::unordered_map<int, T>& dict, const char delimiter) {
+    std::stringstream str_buf;
```
Should this be wrapped?
```
Common::C_stringstream(str_buf);
```

StrikerRUS reviewed (commented):
@shiyu1994 This PR is quite large. Don't you mind I'll be providing my review in pieces? Please consider checking the first part below:

shiyu1994 commented:
> @shiyu1994 This PR is quite large. Don't you mind I'll be providing my review in pieces? Please consider checking the first part below:

Ok. I'll check the review in pieces as you go on. Thanks for your time!

shiyu1994 commented:
@StrikerRUS Done with existing comments.

StrikerRUS commented on include/LightGBM/config.h:
```diff
@@ -1013,6 +1013,34 @@ struct Config {
 
   #pragma endregion
 
+  #pragma region Category Encoder Parameters
```
Could you please add tests for that category encoders work with or document if they don't
- `Dataset.add_features_from()` function
- `Dataset.set_feature_name()` function
- `feature_fraction` param
- `feature_fraction_bynode` param
- `monotone_constraints`, `feature_contri`, `forcedsplits_filename`, `interaction_constraints` and other params where user should specify all features in order or something like that (feature number/name)

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -332,6 +332,133 @@ def check_asserts(data):
+        tmp_dataset = str(tmp_path / 'category_encoding_{}_temp_dataset.bin'.format(model_prefix))
+        train_data_1.save_binary(tmp_dataset)
+
+        # checks that Dataset with category_encoders can be saved to and load from file
```
Seems this comment should be 3 lines upper (before `tmp_dataset = `).

shiyu1994 replied:
Resolved via a25f323

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -332,6 +332,133 @@ def check_asserts(data):
+        expected_num_features = X_train.shape[1] + 3 * len(categorical_feature)
+
+        params.update({"categorical_feature": categorical_feature})
+        booster_1 = lgb.train(params, train_data_1, valid_sets=[valid_data_1], valid_names=["valid_data"])
```
Please do something meaningful with `valid_sets` here and in tests below, right now valid set is unused.

shiyu1994 replied:
Resolved via a25f323

shiyu1994 replied:
The `eval` method of the trained booster is called. And evaluation results with different input format of training data are compared.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -332,6 +332,133 @@ def check_asserts(data):
+        expected_num_features = X_train.shape[1] + 3 * len(categorical_feature)
+
+        params.update({"categorical_feature": categorical_feature})
+        booster_1 = lgb.train(params, train_data_1, valid_sets=[valid_data_1], valid_names=["valid_data"])
```
Please check valid data creation via multiple ways: `create_valid`, `set_reference`, `reference` param in constructor, `add_valid`.

shiyu1994 replied:
Resolved via a25f323. All 4 methods are tested in the `test_category_encoding_inner` function.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -332,6 +332,133 @@ def check_asserts(data):
+
+    params = {
+        "objective": "binary",
+        "metric": "auc",
```
Assert for this metric.

shiyu1994 replied:
Resolved via a25f323 by comparing the evaluation results.

StrikerRUS commented on tests/python_package_test/test_basic.py:
```diff
@@ -332,6 +332,133 @@ def check_asserts(data):
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    params = {
+        'objective': 'multiclass',
+        'metric': 'multi_logloss',
```
Assert for this metric.

shiyu1994 replied:
Resolved via a25f323 by comparing the evaluation results.

StrikerRUS reviewed (commented):
Sorry, I'm quite slow in providing a review even in pieces.

shiyu1994 commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,645 @@
+    }
+
+    std::string Name() const override {
+      std::stringstream str_stream;
```
Resolved via a25f323

shiyu1994 commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,96 @@
+/*!
+ * Copyright (c) 2016 Microsoft Corporation. All rights reserved.
```
Resolved via a25f323

shiyu1994 commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,96 @@
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#ifndef LIGHTGBM_PARSER_H_
+#define LIGHTGBM_PARSER_H_
```
Resolved via a25f323

shiyu1994 commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,96 @@
+
+}  // namespace LightGBM
+
+#endif   // LightGBM_PARSER_H_
```
Resolved via a25f323

shiyu1994 commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,96 @@
+
+  /*!
+  * \brief Create an object of parser, will auto choose the format depend on file
+  * \param filename One Filename of data
```
Resolved via a25f323

shiyu1994 commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,96 @@
+#ifndef C_API_DTYPE_INT64
+#define C_API_DTYPE_INT64   (3)  /*!< \brief int64. */
+#endif
+
```
Resolved via a25f323

shiyu1994 replied:
Now `#include <LightGBM/c_api.h>` is included in `parser.cpp`

shiyu1994 commented on include/LightGBM/utils/common.h:
```diff
@@ -1227,6 +1227,49 @@ inline static std::string ArrayToString(const std::vector<T>& arr, size_t n) {
 }
+
+/*!
+* Converts an unordered_map to a string with with key-value pairs separated by the specified delimiter.
```
Resolved via a25f323

shiyu1994 commented on src/boosting/gbdt_model_text.cpp:
```diff
@@ -41,6 +41,13 @@ std::string GBDT::DumpModel(int start_iteration, int num_iteration, int feature_
 
+  if (category_encoding_provider_ != nullptr) {
+    str_buf << "\"category_encoding_provider\":"
+            << "\"" << category_encoding_provider_->DumpModelInfo() << "\"," << '\n';
```
Resolved via a25f323. I've rewrite the part for `CategoryEncodingProvider` in the model file to make it more human readable. Following is an example of how `CategoryEncodingProvider` is saved when `save_model` is called. (`DumpToString` method of `CategoryEncodingProvider` will be called)
```
keep_raw_cat_method=1
num_original_features=30
num_total_features=75
prior_weight=1
num_categorical_features=15
categorical_feature_index=0
num_categorical_feature_values=22
categorical_feature_info=9:28:28 18:0:18 10:34:35 17:1:25 22:0:1 8:11:11 6:1:1 23:0:5 27:0:2 21:0:7 28:0:1 11:66:71 24:0:2 7:3:3 25:0:2 14:35:54 20:0:19 16:4:21 12:75:81 15:4:29 19:0:23 13:56:73
categorical_feature_index=1
num_categorical_feature_values=25
categorical_feature_info=24:7:18 16:35:40 26:1:10 22:8:27 20:17:40 18:38:55 25:6:14 23:5:27 19:23:42 12:14:14 29:3:4 28:6:9 11:3:4 30:1:2 13:22:22 39:0:1 10:5:6 27:5:9 33:1:2 9:1:1 21:18:46 32:0:1 15:33:39 14:32:34 17:34:45
categorical_feature_index=2
num_categorical_feature_values=104
categorical_feature_info=125:0:1 98:1:4 118:0:3 141:0:1 62:3:3 101:0:2 120:0:4 130:0:4 114:1:4 77:10:12 112:0:2 116:0:2 142:0:2 63:3:3 102:2:10 131:0:2 94:8:13 127:0:4 48:1:1 82:16:18 124:0:2 87:13:17 108:1:4 71:9:10 72:4:4 109:0:5 115:0:3 78:17:18 64:5:5 143:0:2 135:0:2 56:2:2 61:8:8 83:6:7 93:1:5 85:9:12 122:0:4 43:1:1 121:0:3 119:0:2 105:0:2 70:12:12 144:0:1 65:5:5 129:0:5 166:0:1 186:0:1 107:0:4 91:2:5 165:0:1 86:7:7 80:6:6 171:0:1 92:5:7 188:0:1 53:1:1 132:0:4 155:0:1 76:12:12 67:3:3 104:2:3 137:0:3 58:3:3 174:0:1 95:4:6 100:0:4 153:0:1 74:8:8 51:1:1 133:0:2 54:4:4 152:0:2 73:15:16 110:0:3 79:9:9 158:0:1 96:5:9 97:4:8 134:0:5 55:2:2 66:5:5 103:0:6 182:0:1 69:3:3 113:0:3 123:0:4 126:0:2 47:2:2 89:5:5 111:0:4 99:1:1 84:12:12 88:15:17 90:2:6 60:5:5 106:1:3 128:0:4 138:0:2 59:7:7 117:0:4 68:8:8 147:0:2 81:10:12 75:11:12
categorical_feature_index=3
num_categorical_feature_values=385
categorical_feature_info=355:1:1 689:1:1 257:1:1 1104:0:1 395:1:1 1311:0:1 602:1:2 580:1:1 542:1:1 280:1:1 948:0:1 557:1:1 579:1:1 1288:0:1 599:1:2 1308:0:1 1250:0:1 541:1:1 1123:0:1 288:2:2 290:2:2 685:1:2 1203:0:1 1148:0:1 651:2:2 1404:0:1 1319:0:1 610:0:1 645:0:1 719:0:1 640:2:2 347:1:1 684:1:2 701:0:1 363:1:1 1878:0:1 1169:0:1 530:1:2 442:2:2 446:2:2 1155:0:1 809:0:1 453:1:1 1162:0:1 469:1:1 980:0:1 271:2:2 506:2:3 503:0:1 475:0:1 880:1:1 546:3:3 476:3:3 813:0:1 467:1:1 807:0:1 470:1:1 396:2:2 334:2:2 671:0:1 181:1:1 389:1:1 509:1:1 230:1:1 904:0:2 758:0:2 496:2:2 404:1:1 1145:0:1 471:1:1 480:2:2 817:0:1 143:1:1 572:1:2 1217:0:1 508:2:2 336:1:1 673:1:1 1347:0:1 1007:0:1 333:1:1 392:1:1 507:3:3 170:1:1 1102:0:1 428:2:2 595:0:1 398:1:1 344:1:1 1088:0:1 587:1:1 250:1:1 552:1:1 1261:0:1 244:2:2 918:0:1 575:3:3 578:0:2 241:1:1 408:2:2 1076:0:2 402:2:2 781:0:1 447:1:1 269:1:1 606:1:1 401:1:1 1110:0:1 1482:0:1 773:0:1 609:1:2 710:0:1 373:2:2 1384:0:1 311:2:2 648:0:1 421:1:1 1130:0:1 793:0:1 295:1:1 632:1:1 1174:0:1 465:1:1 371:1:2 1206:0:1 869:0:1 201:1:1 1138:0:2 464:3:3 716:0:2 584:0:1 417:1:1 380:1:1 338:1:1 493:2:2 534:0:1 437:1:1 279:1:1 1290:0:1 616:1:1 1230:0:1 519:1:2 2501:0:1 455:2:2 1761:0:1 736:0:1 399:2:2 815:0:1 1152:0:1 593:1:1 912:0:1 203:1:1 1214:0:1 704:0:1 1041:0:1 529:1:1 1274:0:1 600:1:1 307:1:1 644:0:1 272:1:1 981:0:1 641:1:1 1024:0:1 857:0:1 520:3:3 690:1:1 1364:0:1 1027:0:1 227:1:1 504:1:1 951:0:1 618:1:1 390:1:1 463:1:1 800:1:1 797:0:1 1407:0:1 698:0:1 361:1:2 531:0:1 1479:0:1 805:0:1 468:1:1 386:2:3 553:1:1 502:1:1 527:1:1 1491:0:1 782:0:2 445:2:2 448:1:1 449:2:2 1167:0:1 458:2:2 1841:0:1 1132:0:1 438:1:1 384:2:2 1386:0:1 712:0:1 766:0:1 432:4:5 499:0:1 666:1:1 1747:0:1 329:1:1 656:2:3 489:2:2 826:0:1 492:1:1 433:1:1 633:1:2 466:4:4 803:0:1 545:0:1 378:2:2 302:1:1 1040:0:1 1075:0:2 366:2:2 646:1:1 611:1:1 1320:0:1 278:1:1 798:0:1 461:1:1 294:1:1 642:0:2 481:2:2 462:2:2 963:0:1 623:1:1 409:2:2 477:2:4 2499:0:1 372:2:2 310:1:1 984:0:1 664:1:2 1001:0:1 289:3:3 998:0:1 1335:0:1 324:1:1 1033:0:1 234:1:1 571:3:3 568:1:1 391:1:1 728:0:1 596:1:1 933:0:1 224:1:1 561:3:3 558:1:1 895:0:1 930:0:1 221:2:2 321:4:4 658:1:2 260:1:1 1306:0:1 597:0:1 928:0:3 991:0:1 317:1:1 657:1:1 320:1:1 674:2:2 1685:0:1 337:1:1 495:1:1 562:1:1 899:0:1 732:0:1 1299:0:1 590:1:1 788:1:1 451:1:1 388:2:2 1264:0:2 555:1:1 514:4:4 1223:0:1 886:0:1 659:1:1 1670:0:1 668:2:3 431:1:2 537:2:2 1686:0:1 268:1:1 1094:0:1 420:3:3 426:2:2 1194:0:1 485:1:1 994:0:1 285:1:1 245:1:1 582:1:2 412:2:2 538:3:3 1247:0:1 573:1:1 457:1:1 1245:0:1 536:1:1 416:1:1 512:4:4 178:1:1 515:1:1 516:2:2 1297:0:1 588:1:2 415:1:1 248:1:1 585:2:2 678:0:1 441:2:2 422:1:1 365:1:1 394:1:1 731:0:1 359:1:1 1068:0:1 858:0:1 357:1:1 1234:0:1 525:1:1 358:1:1 982:0:1 273:1:1 947:0:1 440:0:1 832:0:1 2250:0:1 761:1:2 603:1:1 293:1:1 513:1:1 680:1:1 346:1:1 748:0:1 403:2:2 1077:0:1 662:1:1 2010:0:1 592:1:1 551:2:2 1260:0:1 992:1:1 491:1:1 566:2:2 381:2:2 1191:0:1 1006:0:1 629:1:2 418:1:1 1092:0:1 686:1:1 349:1:1 559:1:2 693:1:1 427:1:1 340:1:1 744:0:2 577:0:1 501:1:1 705:0:1 713:0:1 819:1:1 955:0:1 246:1:1 920:0:1 747:1:1 1682:0:1 264:1:1 682:0:1 523:1:1 370:1:1 406:2:2 360:1:1 840:0:1
categorical_feature_index=4
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=5
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=6
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=7
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=8
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=9
num_categorical_feature_values=1
categorical_feature_info=0:318:512
categorical_feature_index=10
num_categorical_feature_values=3
categorical_feature_info=1:0:17 2:0:2 0:318:493
categorical_feature_index=11
num_categorical_feature_values=5
categorical_feature_info=2:23:34 1:165:277 0:128:197 3:1:3 4:1:1
categorical_feature_index=12
num_categorical_feature_values=14
categorical_feature_info=0:10:10 7:0:12 3:25:65 10:0:1 5:2:21 1:175:192 18:0:1 6:0:7 8:0:7 11:0:1 9:0:3 2:101:152 4:5:39 21:0:1
categorical_feature_index=13
num_categorical_feature_values=115
categorical_feature_info=56:0:1 40:0:4 75:0:1 50:1:2 158:0:1 61:0:1 89:0:1 10:5:5 32:3:6 111:0:2 7:2:2 34:4:6 96:0:1 17:22:22 101:0:1 22:10:14 180:0:1 13:12:13 69:0:2 138:0:1 27:8:10 11:14:14 28:10:12 67:0:5 54:0:4 81:0:3 57:0:2 6:1:1 35:1:4 30:7:7 109:0:1 137:0:1 68:0:3 31:3:8 38:0:2 46:1:1 80:0:2 43:1:5 122:0:1 47:0:1 134:0:1 60:0:2 139:0:1 99:0:1 20:24:24 90:0:2 164:0:1 156:0:1 77:1:2 124:0:1 45:0:6 88:0:1 102:0:2 23:14:17 98:0:1 19:21:26 83:0:2 120:0:1 106:0:1 92:0:1 95:0:1 16:17:18 58:0:3 44:1:5 155:0:1 76:0:1 118:0:1 39:3:5 63:0:1 100:0:1 21:14:15 104:0:2 25:7:9 133:0:1 59:0:1 42:1:1 79:0:1 153:0:2 74:0:1 48:1:4 66:0:1 233:0:1 37:1:1 116:0:2 71:0:2 14:14:15 93:0:3 51:0:1 130:0:2 53:1:3 15:14:14 94:0:2 52:0:3 33:3:7 112:0:1 70:0:2 18:20:23 97:0:2 9:9:9 176:0:1 12:15:15 49:1:5 128:0:1 115:0:1 36:1:4 41:1:4 542:0:1 525:0:1 24:9:14 103:0:2 105:0:1 26:8:9 8:5:5 87:0:3 29:7:9
categorical_feature_index=14
num_categorical_feature_values=1
categorical_feature_info=0:318:512
num_category_encoders=3
category_encoder=0
type=target_encoder_label_mean
prior=0.621094
categorical_feature_index_to_encoded_feature_index=5:35 4:34 3:33 2:32 0:30 1:31 6:36 7:37 8:38 13:43 14:44 12:42 11:41 10:40 9:39
category_encoder=1
type=count_encoder
categorical_feature_index_to_encoded_feature_index=5:50 4:49 3:48 2:47 0:45 1:46 6:51 7:52 8:53 13:58 14:59 12:57 11:56 10:55 9:54
category_encoder=2
type=target_encoder
prior=0.5
categorical_feature_index_to_encoded_feature_index=5:65 4:64 3:63 2:62 0:60 1:61 6:66 7:67 8:68 13:73 14:74 12:72 11:71 10:70 9:69
```
And for JSON, I've add a separate method `DumpToJSON` for `CategoryEncodingProvider`, which uses `json11` library to gather the information needed for a model from `CategoryEncodingProvider` into a `Json` object. And then dump it to a string. So it should produce reasonable JSON format now. Here's a segment from the dumped JSON string. (The whole is too long to be displayed here.)
```
'category_encoding_provider': {
    'category_encoders': [
        {'categorical_feature_index_to_encoded_feature_index': [{'cat_fid': 1, 'convert_fid': 31}, {'cat_fid': 0, 'convert_fid': 30}, {'cat_fid': 2, 'convert_fid': 32}, {'cat_fid': 3, 'convert_fid': 33}, {'cat_fid': 5, 'convert_fid': 35}, {'cat_fid': 4, 'convert_fid': 34}, {'cat_fid': 6, 'convert_fid': 36}, {'cat_fid': 7, 'convert_fid': 37}, {'cat_fid': 8, 'convert_fid': 38}, {'cat_fid': 10, 'convert_fid': 40}, {'cat_fid': 9, 'convert_fid': 39}, {'cat_fid': 11, 'convert_fid': 41}, {'cat_fid': 12, 'convert_fid': 42}, {'cat_fid': 14, 'convert_fid': 44}, {'cat_fid': 13, 'convert_fid': 43}], 'name': 'target_encoder', 'prior': 0.621094}, 
        {'categorical_feature_index_to_encoded_feature_index': [{'cat_fid': 1, 'convert_fid': 46}, {'cat_fid': 0, 'convert_fid': 45}, {'cat_fid': 2, 'convert_fid': 47}, {'cat_fid': 3, 'convert_fid': 48}, {'cat_fid': 5, 'convert_fid': 50}, {'cat_fid': 4, 'convert_fid': 49}, {'cat_fid': 6, 'convert_fid': 51}, {'cat_fid': 7, 'convert_fid': 52}, {'cat_fid': 8, 'convert_fid': 53}, {'cat_fid': 10, 'convert_fid': 55}, {'cat_fid': 9, 'convert_fid': 54}, {'cat_fid': 11, 'convert_fid': 56}, {'cat_fid': 12, 'convert_fid': 57}, {'cat_fid': 14, 'convert_fid': 59}, {'cat_fid': 13, 'convert_fid': 58}], 'name': 'count_encoder'}, 
        {'categorical_feature_index_to_encoded_feature_index': [{'cat_fid': 1, 'convert_fid': 61}, {'cat_fid': 0, 'convert_fid': 60}, {'cat_fid': 2, 'convert_fid': 62}, {'cat_fid': 3, 'convert_fid': 63}, {'cat_fid': 5, 'convert_fid': 65}, {'cat_fid': 4, 'convert_fid': 64}, {'cat_fid': 6, 'convert_fid': 66}, {'cat_fid': 7, 'convert_fid': 67}, {'cat_fid': 8, 'convert_fid': 68}, {'cat_fid': 10, 'convert_fid': 70}, {'cat_fid': 9, 'convert_fid': 69}, {'cat_fid': 11, 'convert_fid': 71}, {'cat_fid': 12, 'convert_fid': 72}, {'cat_fid': 14, 'convert_fid': 74}, {'cat_fid': 13, 'convert_fid': 73}], 'name': 'target_encoder', 'prior': 0.5}
    ], 
    'category_infos': [
        {'categorical_feature_index': 0, 'label_and_count_info': [{'categorical_feature_value': 16, 'count': 21, 'label_sum': 4}, {'categorical_feature_value': 20, 'count': 19, 'label_sum': 0}, {'categorical_feature_value': 14, 'count': 54, 'label_sum': 35}, {'categorical_feature_value': 11, 'count': 71, 'label_sum': 66}, {'categorical_feature_value': 8, 'count': 11, 'label_sum': 11}, {'categorical_feature_value': 25, 'count': 2, 'label_sum': 0}, {'categorical_feature_value': 15, 'count': 29, 'label_sum': 4}, {'categorical_feature_value': 12, 'count': 81, 'label_sum': 75}, {'categorical_feature_value': 19, 'count': 23, 'label_sum': 0}, {'categorical_feature_value': 0, 'count': 2, 'label_sum': 0}, {'categorical_feature_value': 17, 'count': 25, 'label_sum': 1}, {'categorical_feature_value': 7, 'count': 3, 'label_sum': 3}, {'categorical_feature_value': 22, 'count': 1, 'label_sum': 0}, {'categorical_feature_value': 10, 'count': 35, 'label_sum': 34}, {'categorical_feature_value': 27, 'count': 2, 'label_sum': 0}, {'categorical_feature_value': 18, 'count': 18, 'label_sum': 0}, {'categorical_feature_value': 6, 'count': 1, 'label_sum': 1}, {'categorical_feature_value': 23, 'count': 5, 'label_sum': 0}, {'categorical_feature_value': 21, 'count': 7, 'label_sum': 0}, {'categorical_feature_value': 28, 'count': 1, 'label_sum': 0}]}, 
        {'categorical_feature_index': 1, 'label_and_count_info': [{'categorical_feature_value': 11, 'count': 4, 'label_sum': 3}, {'categorical_feature_value': 28, 'count': 9, 'label_sum': 6}, {'categorical_feature_value': 13, 'count': 22, 'label_sum': 22}, {'categorical_feature_value': 30, 'count': 2, 'label_sum': 1}, 
...
```

shiyu1994 commented on src/c_api.cpp:
```diff
@@ -862,24 +874,8 @@ std::function<std::vector<std::pair<int, double>>(T idx)>
-  bool is_end_ = false;
-  std::function<std::pair<int, double>(int idx)> iter_fun_;
-};
+std::function<float(int row_idx)>
```
Resolved via a25f323

shiyu1994 commented on src/c_api.cpp:
```diff
@@ -1290,6 +1337,32 @@ int LGBM_DatasetCreateFromCSC(const void* col_ptr,
   int32_t nrow = static_cast<int32_t>(num_row);
+
+  std::vector<std::unique_ptr<CSC_RowIterator>> csc_iterators(ncol_ptr - 1);
+  #pragma omp parallel for schedule(static)
```
Resolved via a25f323

shiyu1994 commented on src/boosting/gbdt_model_text.cpp:
```diff
@@ -478,6 +491,14 @@ bool GBDT::LoadModelFromString(const char* buffer, size_t len) {
+  } else {
+    Log::Fatal("Model file doesn't specify category_encoding_provider");
+    return false;
+  }
```
Increment model version to `v4` and support backward compatibility of reading old model, via 66826a2.

shiyu1994 commented:
@StrikerRUS Thanks for your detailed review. I've resolved most of comments, except the compatibility with methods like `add_features_from`, and the two `TODO` issues. I'll do these in further commits.

shiyu1994 commented on src/c_api.cpp:
```diff
@@ -1195,6 +1238,7 @@ int LGBM_DatasetCreateFromCSR(const void* indptr,
   API_END();
 }
 
+// TODO(shiyu1994): This is called in mmlspark, supporting category encoding mode with this function requires modifying mmlspark.
```
Can we ignore `category_encoders` when a dataset is constructed from `LGBM_DatasetCreateFromCSRFunc` for now? 
This method is not used by our python or R-package. It probably involves corresponding modifications of related  external projects who use this function to enable `category_encoders`. Maybe we can leave the modification in future PR? WDYT?

shiyu1994 commented on src/c_api.cpp:
```diff
@@ -934,10 +937,11 @@ int LGBM_DatasetCreateFromSampledColumn(double** sample_data,
   }
-  DatasetLoader loader(config, nullptr, 1, nullptr);
+  DatasetLoader loader(&config, nullptr, 1, nullptr);
+  // TODO(shiyu1994): in this case do we need category encoding provider ? it seems that therea are no labels available.
```
Same problem as https://github.com/microsoft/LightGBM/pull/3234#discussion_r740238415

shiyu1994 commented:
@jameslamb @StrikerRUS @guolinke Could you please take a look at this PR again? We plan to merge this recently.

BTW, I have an idea to keep both the old and new interfaces for data construction in `src/c_api.cpp`. This can avoid unexpectedly breaking other projects leveraging these C APIs. WDYT?

StrikerRUS commented:
> I have an idea to keep both the old and new interfaces for data construction in `src/c_api.cpp`. This can avoid unexpectedly breaking other projects leveraging these C APIs. WDYT?

I believe this would be great!

hzy46 commented on docs/Advanced-Topics.rst:
```diff
@@ -15,22 +15,50 @@ Missing Value Handle
+   -  ``count`` to indicate count encoding in method 2.
+
+   Note that the aforementioned methods can be used simultaneously. Different methods are separated by commas.
+   For example ``category_encoders=target:0.5,target:count,raw`` will enable using splits with method 1, and in addition, convert each categorical feature into 3 numerical features. The first one uses target encoding with prior ``0.5``. The second one uses target encoding with default prior, which is the mean of labels of the training data. The third one uses count encoding.
```
I'm a bit confused here. Should we change ``category_encoders=target:0.5,target:count,raw``  to ``category_encoders=target:0.5,target,count,raw``?

shiyu1994 replied:
Oh, yes. I did not notice that. Thanks!

shiyu1994 replied:
Resolved via a4ec01f

shiyu1994 commented:
Gently ping @jameslamb @StrikerRUS @guolinke I've kept the old C APIs with the latest commits. Could you please check whether your previous comments have been addressed? Thanks.

shiyu1994 commented:
/gha run r-valgrind

shiyu1994 commented:
/gha run r-solaris

shiyu1994 commented:
/gha run r-solaris

guolinke commented:
@shiyu1994 
are there any new benchmark results?
And is ensemble of encoders used or not?

shiyu1994 commented:
@guolinke There's no updates in the encoding methods since the previous benchmarks. Ensemble is not used in the benchmark results listed in this PR. Benchmark results with ensemble trick can be found in https://github.com/microsoft/LightGBM/issues/4779.

shiyu1994 commented:
/gha run r-solaris

shiyu1994 commented:
The Solaris CRAN Check failed again. It is weird that the compiler reports an error for `label_t`
```
c_api.cpp:860:17: error: ‘label_t’ is already declared in this scope
 using LightGBM::label_t;
                 ^
```
But not for `using LightGBM::data_size_t;` in several lines above in `c_api.cpp`. Both `data_size_t` and `label_t` are defined in `LightGBM/meta.h`.

shiyu1994 commented:
According to https://github.com/collectd/collectd/issues/1301, it seems that Solaris system header also defines a `label_t` type, which causes the conflict. In that case, I'll use `LightGBM::label_t` in `c_api.cpp`.

shiyu1994 commented:
/gha run r-solaris

shiyu1994 commented:
/gha run r-solaris

shiyu1994 commented:
/gha run r-valgrind

shiyu1994 commented:
We finally get all the tests passed. But it seems that the status of `Solaris CRAN check` is not updated.

jameslamb commented:
/gha run r-solaris

jameslamb commented:
> According to [collectd/collectd#1301](https://github.com/collectd/collectd/issues/1301), it seems that Solaris system header also defines a `label_t` type, which causes the conflict. In that case, I'll use `LightGBM::label_t` explicitly in `c_api.cpp`.

thanks for looking into this and for finding a fix! Not surer why the status for Solaris didn't update. I just ran it again.

jameslamb commented on R-package/tests/testthat/test_basic.R:
```diff
@@ -2079,7 +2079,8 @@ test_that("lgb.train() works with linear learners when Dataset has categorical f
     , num_leaves = 2L
-    , categorical_feature = 1L
+    # indices in categorical feature start from 1L, so here should be 2L
+    , categorical_feature = 2L
```
similar to my previous comment...is this absolutely necessary for this PR? Or could it be separate?

shiyu1994 replied:
Resolved via fab88db. Remove this change. And will open a separate issue for this.

jameslamb commented on docs/Advanced-Topics.rst:
```diff
@@ -15,22 +15,50 @@ Missing Value Handle
+
+   -  Method 2: Encoding categorical features into numerical values. We provide two encoding options:
+
+      -  **Target encoding**: encode the categorical feature value by the mean of labels of data with the same feature value in the training set. It is easy to overfit the training data if the encoded value of a training data point uses the label of that training data point itself. So we would randomly divide the training data into folds, and when calculating the target encoding for data in one f ...
```
```suggestion
      -  **Target encoding**: encode the categorical feature value by the mean of labels of data with the same feature value in the training set. It is easy to overfit the training data if the encoded value of a training data point uses the label of that training data point itself. So LightGBM randomly divides the training data into folds, and when calculating the target encoding for data in one fold, only considers data in other folds.
```

I think it's important to remove language like "I" or "we" or "you" from documentation talking about LightGBM's behavior, to avoid users being confused and thinking that this doc means they need to do some preprocessing themselves before passing data to LightGBM.

shiyu1994 replied:
Resolved via fab88db.

jameslamb commented on R-package/tests/testthat/test_basic.R:
```diff
@@ -2207,6 +2208,96 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
+    , valids = list("valid1" = dtest)
+  )
+  pred3 <- bst$predict(test$data)
+  expect_equal(dim(dtrain), c(6513L, 134L))
```
am I right here the the purpose of this test on `dim()` is to check that additional features were added based on the values passed to `category_encoders`?

The `lightgbm::agaricus` dataset has 6513 rows and 126 features, so I guess this is saying "with `target,count,raw`", LightGBM should add 8 new features...one `count` feature and one `target`  feature per categorical feature.

If I'm right about that, then please consider this change to the tests to make it more obvious that that is what is happening here. Otherwise, this might fail some day and we won't know what the hard-coded numbers 6513 and  134 were supposed to mean.

```suggestion
  # one new "count" and "target" feature is added per categorical feature
  num_new_cat_features <- length(cat_fid) * 2L
  expect_equal(dim(dtrain), c(nrow(train$data), ncol(train$data) + num_new_cat_features))
```

shiyu1994 replied:
Resolved via fab88db.

jameslamb reviewed (changes requested):
I took a first look today and left a few comments. It's a very large PR so apologies, it will take a while for me to give a thorough review.

shiyu1994 commented on R-package/tests/testthat/test_basic.R:
```diff
@@ -2207,6 +2208,96 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
 
 })
 
+test_that(paste0("Category encoding for R package works"), {
```
Resolved via fab88db

shiyu1994 commented on R-package/tests/testthat/test_basic.R:
```diff
@@ -2207,6 +2208,96 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
+  dtrain <- lgb.Dataset(train$data, label = train$label,
+    categorical_feature = cat_fid, category_encoders = "raw")
+  dtest <- lgb.Dataset(test$data, label = test$label,
+    categorical_feature = cat_fid, reference = dtrain)
```
Resolved via fab88db

shiyu1994 commented on R-package/tests/testthat/test_basic.R:
```diff
@@ -2207,6 +2208,96 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
+  dtrain <- lgb.Dataset(train$data, label = train$label,
+    categorical_feature = cat_fid)
+  dtest <- lgb.Dataset(test$data, label = test$label,
+    categorical_feature = cat_fid, reference = dtrain)
```
Resolved via fab88db

shiyu1994 commented on R-package/R/lgb.Dataset.R:
```diff
@@ -641,6 +651,15 @@ Dataset <- R6::R6Class(
+        # categorical features from the parameters and directly set it in R.
+        self$set_categorical_feature(params$categorical_feature)
+        params$categorical_feature <- NULL
+      }
```
Resolved via fab88db. Remove this change. And will open a separate issue for this.

shiyu1994 commented:
@jameslamb Thank you! I've addressed your suggestions so far. And a separate issue or pull request should be open for this https://github.com/microsoft/LightGBM/pull/3234#discussion_r750359184

shiyu1994 commented:
Gently ping @guolinke @jameslamb @StrikerRUS again for a look of this PR. Hope that we can move faster towards 4.0.0. Thank you!

guolinke reviewed (approved):
Thank you @shiyu1994 , this PR overall looks good to me.

jameslamb commented:
> Gently ping again for a look of this PR.

Can you please update this branch to latest `master`? Then we can run the additional CI jobs like valgrind and Solaris checks.

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+// transform categorical features to encoded numerical values before the bin construction process
+class CategoryEncodingProvider {
+ public:
+  class CatConverter {
```
Can we put CatConverter as separate class in separate file to reduce complexity for CategoryEncodingProvider? Looks like we might need to access properties in CategoryEncodingProvider, then we can pass CategoryEncodingProvider to the CatConverter.

tongwu-msft replied:
For every converter we can have separate unit tests?

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+
+    virtual std::string DumpToString() const = 0;
+
+    virtual json11::Json DumpToJSONObject() const = 0;
```
To make sure we have common pattern for serialization could we use json as general pattern? json is a special string then we do not need to define our customize pattern?

tongwu-msft replied:
And I believe we should not do serialization & deserialization very often?

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+      return cat_fid_to_convert_fid_.at(cat_fid);
+    }
+
+    static CatConverter* CreateFromCharPointer(const char* char_pointer, size_t* used_len, double prior_weight) {
```
we can have separate file for class initialization logic.

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+      return str_stream.str();
+    }
+
+    json11::Json DumpToJSONObject() const override {
```
Looks like it is another serialization logic for sub/base class, could we have pattern like below: 

DumpToJSONObject(json11::Json context)
{
     base.DumpToJSONObject(context);

     // Dump to json logic for sub class
}

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+    }
+  };
+
+  class TargetEncoderLabelMean: public CatConverter {
```
For these converter, looks like separate class would be clear?

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+  }
+
+  // parameter configuration
+  Config config_;
```
Looks like we should not have config after initialize?

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+  Config config_;
+
+  // size of training data
+  data_size_t num_data_;
```
for below properties we'd better have only 1 unified pair serialization & deserialization function?

tongwu-msft commented on include/LightGBM/category_encoding_provider.hpp:
```diff
@@ -0,0 +1,771 @@
+  bool accumulated_from_file_;
+};
+
+class CategoryEncodingParser : public Parser {
```
Use parser as middleware looks like wield, it is more like a step in pipeline processing. Not sure is that possible we refactor the processing steps to pipeline pattern?

tongwu-msft commented on include/LightGBM/parser_base.h:
```diff
@@ -0,0 +1,139 @@
+  /*!
+  * \brief Constructor for customized parser. The constructor accepts content not path because need to save/load the config along with model string
+  */
+  explicit Parser(std::string) {}
```
to have a constructor of string looks like wield, not sure any purpose for this in base class?

tongwu-msft commented on src/io/category_encoding_provider.cpp:
```diff
@@ -0,0 +1,1330 @@
+
+namespace LightGBM {
+
+CategoryEncodingProvider::CategoryEncodingProvider(Config* config) {
```
Should we have a unified construction for CategoryEncodingProvider? And others are just transform inputs into this construction method? like below:

CategoryEncodingProvider(a, b, c, d, e)

CategoryEncodingProvider(a, b, c, d) {
CategoryEncodingProvider((a, b, c, d, 0)
}

tongwu-msft commented on src/io/category_encoding_provider.cpp:
```diff
@@ -0,0 +1,1330 @@
+  tmp_parser_ = nullptr;
+}
+
+std::string CategoryEncodingProvider::DumpToJSON() const {
```
Same as converter, Looks like unified single pair serialization/deserialization would reduce lots of complexity.

tongwu-msft commented on src/io/category_encoding_provider.cpp:
```diff
@@ -0,0 +1,1330 @@
+  }
+}
+
+void CategoryEncodingProvider::SyncEncodingStat(std::vector<std::unordered_map<int, double>>* fold_label_sum_ptr,
```
Not quite understand this SyncEncodingStat method. Not sure we can add more comments for it?

shiyu1994 replied:
Resoved via 3e91f68

shiyu1994 commented:
Thanks @tongwu-msft for the valuable suggestions for refactoring the code. We can have follow-up PRs for the refactor after this is merged.

StrikerRUS commented:
> We can have follow-up PRs for the refactor after this is merged.

TBH, I'm not sure this is a good approach of adding new features for the open source project where there are no any deadlines and the work is based on a collaboration of multiple persons. We may end up having a lot of such "valuable suggestions for refactoring the code" that may be never get enough attention in the future. As a result, the whole codebase of LightGBM will become less readable, unoptimized and extremely hard to make contributions for outside contributors.

shiyu1994 commented:
@StrikerRUS Thanks for the reminder. I've discussed with @tongwu-msft, and he will do the refactor and then push to this branch.

jameslamb commented:
I agree with @StrikerRUS 's comments in https://github.com/microsoft/LightGBM/pull/3234#issuecomment-987869466. Right now, while attention is focused on this PR, is the best time to address review comments.

guolinke commented:
@shiyu1994 should we close this PR? I think this is hard to merge now.

jameslamb commented:
It's been 6 months since @guolinke asked if we should close this PR with no response, and more than a year since the most recent commit.

@shiyu1994 I'm closing this. If you'd like to continue this work, please propose a new PR and we can start a new review cycle.
